src/docs/binding-generics.md

```diff
@@ -0,0 +1,189 @@
+# Binding generics
+
+This document explores projections of Swift generics into C#.
+
+1. [Generic functions](#generic-functions)
+2. Generic types (TBD)
+3. Generics with protocol constraints (TBD)
+4. Generics with PAT constraints (TBD)
+
+## Generic functions
+
+When a generic type is used in Swift, its memory layout isn't known at compile time. To manage this, the method parameter whose type is generic represented by an opaque pointer. To handle this:
+
+- **Opaque Pointers**: Swift represents generic parameters as opaque pointers (`%swift.opaque`), abstracting the actual memory layout.
+- **Type Metadata**: Swift includes a type metadata pointer (`%T`) as an implicit argument at the end of a function's signature. This metadata provides the runtime information needed to manage the generic type.
+- **Indirect Return**: For returning generic types, Swift uses an [indirect return result](https://github.com/dotnet/runtime/issues/100543), where the caller provides a memory location for the return value.
+
+Consider the following simple swift function:
+
+```swift
+func returnData<T>(data: T) -> T {
+    return data
+}
+```
+
+When compiled into LLVM-IR the above function has the following LLVM-IR signature:
+
+```llvm
+@"output.returnData<A>(data: A) -> A"(ptr noalias nocapture sret(%swift.opaque) %0, ptr noalias nocapture %1, ptr %T)
+```
+
+- `%0`: Indicates an **indirect return result**, where the caller allocates memory for the return value.
+- `%1`: The `data` parameter, passed as a pointer.
+- `%T`: The **type metadata** for `T`.
+
+### Projecting into C\#
+
+A sample projection into C# assuming that the type argument is projected as struct could look like this:
+
+```csharp
+[DllImport(Path, EntryPoint = "...")]
+[UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+public static extern void PInvoke_ReturnData(SwiftIndirectResult result, void* data, void* metadata);
```

Comment:
With regards to getting type metadata, I think that the shape of the accessor should look like this:
```csharp
public interface INominalType {
    static TypeMetadata TypeMetadata { get; } // or TypeMetadata GetTypeMetadata ();
}

public class SomeBoundGenericType<T> : INominalType {
     static TypeMetadata _tmd PinvokesForSomeBoundGenericType.MetadataAccessor(MetadataRequest.Complete, Runtime.GetMetadataOrThrow(typeof (T)));
    TypeMetadata INominalType.TypeMetadata => _tmd;
}
internal static class PInvokesForSomeBoundGenericType {
    [DllImport(/* ... */)]
    public static extern TypeMetadata MetadataAccessor(MetadataRequest rq, TypeMetadata t0);
```
Runtime.GetMetadata can have a number of forms:
```
public static TypeMetadata GetMetadata(Type t) { } // what happens on fail?
// and/or
public static bool TryGetTypeMetadata(Type t, [NotNullWhen(true)]out TypeMetadata? md) { }
// and/or
public static TypeMetadata GetMetadataOrThrow(Type t)
{
    if (TryGetTypeMetadata(t, out var md)) {
        return md;
    } else {
        throw new RuntimeException($"Unable to get metadata for {t.Name}");
   }
}
```
I'd like to see the implementation of `TryGetTypeMetadata` to include a case along the lines of:
```csharp
    if (typeof(INominalType).IsAssignableFrom (t)) {
        var pi = t.GetProperty("INominalType.TypeMetadata", BindingFlags.Static);
        md = (TypeMetadata)pi.GetValue(null)!;
        return true;
    }
```

Reply:
Ideally we would avoid reflection as much as possible. I know that for generics this is a bit tricky as they're "unknown" by design, but maybe we can rely on C# generics in that case. 

Theoretically speaking we should not need reflection anywhere - when we generate projections we know of all possible types which will occur since we have to generate projections for them. So we may be able to rely on a generated lookup table instead of on reflection.

Reflection will get tricky due to AOT/Trim compatibility. The example above is not trim compatible and AFAIK there's no way to make it really compatible since we can't put annotation on the `t` as it's completely unknown. Also I would be careful about calling static interface methods through reflection - we need to try it, there were runtime bugs around this even on CoreCLR.

Reply:
Actually the trimability of this approach is something I'm not sure about. The trimming rules for static interface methods are rather complicated - and I don't know how it interacts with reflection. We will need to ask...
In any case I would expect the above to produce a Trim warning.

Reply:
Just a wild idea. Maybe we could replace reflection with some usage of source generators. For example for primitive types like `Int` we could define a `struct Swift.IntWrapper : ISwiftObject<Swift.IntWrapper>`. The interface would provide us methods to get metadata etc. Then in order to allow the users to call the generics using regular C# `Int` we would have a source generator which would take their function call and generate a function which wraps the primitive types and calls the correct internal generic function. This way the user would get compile-time errors when trying to call a generic with unsupported argument.

Reply:
I thought about this some more... and I have to admit small amount of reflection is probably better here. We could also use reflection to create a `Helper<T>` over the `t` with a constraint and inside just call the static interface method - not sure which one is better, both are probably currently problematic for AOT/trim.

That said, this is basically an equivalent code to calling an instance interface method on an `object` - and that is something trimming does support (and should) even if there are no annotations anywhere. The static method is problematic because there's no C# syntax for the instance equivalence:
```
void MYTest(object o)
{
    if (o is ISwiftObject so)
    {
        so.GetMetadata();
    }
}
```

C# can only do this through generics, but it can't do it on `Type`. So it might mean we need a new trimming feature or some kind of interesting workaround, but it should be possible without annotating `t`. Trimmer (and AOT) already works by the rule that "If the interface method is used and the type which implements it is used, then the type's implementation of that method must be kept" - static interface methods should in general work in a similar way (and if not we would have to fix the trimming).

Source generators in this case are a bit problematic - they only work well within one assembly. So for example, we can't rely on the source generator to run on a type in some random assembly - we would have to tied the usage of that somewhere. Another problem is that you might end up with 10 different generated implementations of `IntWrapper` - for value types this is probably OK (just wasteful), but if we need object identity it might become a problem. There are also versioning considerations.

And replacing reflection with a big lookup table has its own issues (how do we initialize the table if we need to merge two generated tables from two different assemblies -> type collisions -> problems - see JsonContext problems in JSON and ASP.NET DI). That said - we will have the problem of type identity across modules anyway, so that's another thing we need to potentially resolve.

Reply:
Why do we need reflection here? Aren't the metadata accessors known at bind time?

Reply:
I understand the issue of trimming. In order to do a fully correct implementation based on static analysis alone requires solving a problem akin to the halting problem. I would prefer to have a 100% correct solution in hand at the outset rather than requiring essentially a post processing step that will only be correct in the cases that are determinable at compile time.

Here's an example. Suppose we have the following in Swift:
```swift
public protocol Named {
    func name() -> String
}
public class Foo<T> : Named {
    private var val: T
    private var name: String
    public init(name: String, val: T) {
        self.val = val
    }
    public func getName() -> String {
        return "\(name)\(val)"
    }
}

public func getNamed () -> Named {
     return Foo<Int> ("Alex", 17)
}
```
When bound into C# we will get an Existential container with a concrete instance of Foo in it. We could represent this will a proxy, but in this particular case we can inspect the type metadata in it and find that it is actually a Foo and we can inspect the generic parts and find that it's an nint. We should therefore present this in C# as a `Foo<T>` rather than as a proxy. In this case we would need to make a C# `Foo<nint>` which would have eluded static analysis.
Similarly, I can write pathological cases in C# that will also elude static analysis.

Reply:
So before I sound too negative about this, I think that static analysis will work to a degree and it can be put in place. So assuming we look at the broad shape of the T we need a Swift type metadata for and have a case that looks in a table that contains a mapping of C# bound generic types that are mapped to a cache of Swift type metadata along with a closure that will generate the type metadata the first time it's accessed.
The important thing to note is that on cache miss, you can fall back to reflection - so we can start with reflection based code and add static analysis as a later optimization with no loss of generality.

Reply:
> Similarly, I can write pathological cases in C# that will also elude static analysis.

If you manage to do that, I would be very interested and it's likely a bug :-)

But you have a correct point in the above. I honestly don't know how we're going to solve that. I suspect we might need to be able to fallback to a wrapper. I assume we will need a wrapper for pretty much every protocol anyway for the cases where the underlying type is not projected to C#, so if the projection is for a type which can't be safely constructed (like Foo<int>), then we would fallback to the general wrapper. This is an interesting problem. I think we may need to discuss this in a wider audience. For non-AOT cases, you're absolutely right that using reflection in this case helps a lot. But for AOT this is problematic.

And given that on iOS AOT is basically a must have.... I'd rather we have an AOT compatible solution which potentially doesn't behave ideally (like creating a wrapper in the above case).

I'd also add that we want this behavior to be predictable and stable - it should not be magic. As a developer I need to experience the exact same behavior between debug and runtime and it remains consistent whether I add some more code to my app or not - and ideally, I should be able to predict the behavior by just looking at the projected code. There will be rough edges for sure, but we should try to minimize those. Maybe we could return a generic wrapper which implements `IDynamicCastableInterface` that could fix this for interfaces. We don't have a way to cast to a different class/struct type though, so it's not a solution for `Foo<int>`.

Reply:
> Here's an example. Suppose we have the following in Swift:

How does the problematic code on the C# side look like? I'm trying to understand why we wouldn't know about `Foo<nint>` at compile time. I would think that if someone calls `getNamed` and then proceeds to handle the returned `Named` as not just `Named` but as a `Foo<nint>`, we would see `Foo<nint>` _somewhere_ at AOT compile time, it wouldn't just suddenly dynamically appear at runtime.

Reply:
I think the question if what should happen if C# calls `getNamed()` and doesn't cast it. It would mean we don't see `Foo<nint>` anywhere, so it doesn't exist and thus we would have to fallback to a protocol wrapper. The thing I'm nervous about is that this effectively changes runtime behavior based on type usage in the app and it's observable as I can then call `GetType().ToString()` on the return value and see if it was indeed `Foo<nint>` or the wrapper. It would also change behavior between Debug/Release.

Maybe we'll have to live with this, I'm not sure...

Reply:
Forgetting the generics for a moment - what's the trimming story for this? If `getNamed` returned a `Named` that was also `SomeFoo`. Would we potentially trim `SomeFoo` if nobody was using it? Or are all Swift types rooted in this design?

Reply:
Great question - and it's basically what I was getting at originally... 
I think the answer right now is "We don't know" and somebody will have to figure it out.

Reply:
If we know about `Foo<nint>` at compile time (in the same way that we know about the non-generic `SomeFoo`) this at least leaves a foot in the door to allow making this AOT-safe. I don't know the shape of a solution and I don't know if we'd need new AOT compiler features or it can be handled by a generator, but it would be a start.

Reply:
In looking at this issue, we have a couple of different problems here:
1. how to optimally trim the output assemblies
2. how to keep to the principal of least astonishment (code should behave as you would expect to the greatest degree)

So in this particular case, we have a function, `getNamed` that returns a protocol, or more precisely an existential container that contains a `Foo<Int>` in the payload, the type metadata for `Foo<Int>` and the protocol witness table for `Named` with respect to `Foo<Int>`.

The simplest solution is to create a C# proxy that implements `INamed` and encapsulates the existential container. This allows `Foo<Int>` to be treated as `INamed` but as nothing else beyond the proxy type. This allows trimming but violates the principal of least astonishment because in Swift I could do this:
```swift
let x = getNamed ()
if let y = x as? Foo<Int> {
    print("It's a Foo<Int>")
}
```csharp
Whereas in C# if I did:
```
var x = TopLevelFunctions.GetNamed ();
if (x is Foo<Int> y) {
    Console.WriteLine ("this will never get reached");
}
```
And in theory we can handle this because we have everything we need in the existential container to project the existential container in C# as `Foo<Int>`: a handle to the Swift instance, and its type metadata, but in order to make this case work we need the ability to look up the type metadata in a runtime database to get `Foo<_>` and crack open the type metadata to extract the generic specialization and look it up in the runtime type database to get `nint` which is enough for us to call a factory on `Foo<nint>` to make one from the handle. This, of course, requires reflection. The one piece that this doesn't address is when the protocol is an extension protocol and the bound type `Foo<_>` won't show that it implements the protocol.

As to whether or not this case will be common, my instincts say yes since Apple is very much in the camp of aspect oriented design rather than OO.

I should also point out that this will also be an issue for any Swift function that returns `some` of a PAT and using `some` is the entire basis of SwiftUI.

The final question for us to answer is which of the two goals are more important: size of the output binary or referential transparency.

We might be able to get the best of both worlds in this case by using the a predicate to determine whether or not a type should be trimmed based on if the type is a Swift binding and if it implements one or more protocols. While that sounds promising, because of the heavy use of aspect oriented programming, everything implements protocols.

Reply:
This is probably for another discussion, but things which I found interesting in your reply:

Extension protocols - if we project them as the implementation class, that class will not have the interface, so we can't really do that. So we have to project them as some kind of wrapper - we could still make the wrapper compare equally to the underlying instance of the projected class, but that doesn't solve the problem with type checks. The corresponding `is Foo` will simply not work in C#. I think this is effectively a language limitation, since .NET/C# doesn't have the notion of extension interfaces.

> which of the two goals are more important: size of the output binary or referential transparency

I don't think we have much of an option on the binary size - we need to make it small. Let's assume that some day in the future we will be able to project all of the foundation libraries. At that time, if I write an app which makes a single use of type from Swift I would get all of those types from foundation libraries... and my app would be big. That doesn't sound like a good solution.

Reply:
I will be merging this PR, I opened an issue which we can use for discussion: https://github.com/dotnet/runtimelab/issues/2794